drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c

```diff
@@ -0,0 +1,1808 @@
+
+	f->fmt.pix.bytesperline = get_bytesperline(f->fmt.pix.width, fmt);
+	f->fmt.pix.field = V4L2_FIELD_NONE;
+	f->fmt.pix.sizeimage =
+		get_sizeimage(f->fmt.pix.bytesperline, f->fmt.pix.width,
+			      f->fmt.pix.height, fmt);
+	return 0;
+}
+
+static int vidioc_enum_fmt(struct file *file, void  *priv,
+			   struct v4l2_fmtdesc *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_fmt_list *fmts = &node->supported_fmts;
+	struct bcm2835_isp_fmt *fmt;
+
+	if (f->index < fmts->num_entries) {
+		/* Format found */
+		fmt = &fmts->list[f->index];
+		f->pixelformat = fmt->fourcc;
+		f->flags = fmt->flags;
+		return 0;
+	}
+
+	return -EINVAL;
+}
+
+static int bcm2835_isp_node_enum_fmt_vid_cap(struct file *file, void  *priv,
+					     struct v4l2_fmtdesc *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+
+	if (!NODE_IS_CAPTURE(node))
+		return -EINVAL;
+
+	return vidioc_enum_fmt(file, priv, f);
+}
+
+static int bcm2835_isp_node_enum_fmt_meta_cap(struct file *file, void  *priv,
+					      struct v4l2_fmtdesc *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+
+	if (!NODE_IS_STATS(node))
+		return -EINVAL;
+
+	return vidioc_enum_fmt(file, priv, f);
+}
+
+static int bcm2835_isp_node_enum_fmt_vid_out(struct file *file, void  *priv,
+					     struct v4l2_fmtdesc *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+
+	if (!NODE_IS_OUTPUT(node))
+		return -EINVAL;
+
+	return vidioc_enum_fmt(file, priv, f);
+}
+
+static int bcm2835_isp_node_try_fmt_vid_cap(struct file *file, void *priv,
+					    struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_fmt *fmt;
+
+	if (!NODE_IS_CAPTURE(node))
+		return -EINVAL;
+
+	return vidioc_try_fmt(node, f, fmt);
+}
+
+static int bcm2835_isp_node_try_fmt_meta_cap(struct file *file, void *priv,
+					     struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_fmt *fmt;
+
+	if (!NODE_IS_STATS(node))
+		return -EINVAL;
+
+	fmt = find_format(f, node);
+	if (!fmt) {
+		f->fmt.meta.dataformat = V4L2_META_FMT_BCM2835_ISP_STATS;
+		fmt = find_format(f, node);
+	}
+	f->fmt.meta.buffersize = get_port_data(node)->minimum_buffer.size;
+
+	return 0;
+}
+
+static int bcm2835_isp_node_try_fmt_vid_out(struct file *file, void *priv,
+					    struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_fmt *fmt;
+
+	if (!NODE_IS_OUTPUT(node))
+		return -EINVAL;
+
+	return vidioc_try_fmt(node, f, fmt);
+}
+
+static int bcm2835_isp_node_s_fmt_vid_cap(struct file *file, void *priv,
+					  struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	int ret;
+
+	ret = bcm2835_isp_node_try_fmt_vid_cap(file, priv, f);
+	if (ret)
+		return ret;
+
+	v4l2_dbg(1, debug, &node_get_dev(node)->v4l2_dev,
+		 "%s: Set capture format for node %s[%d]\n",
+		 __func__, node->name, node->id);
+
+	return populate_qdata_fmt(f, node);
+}
```

Comment:
And once more. Disable the invalid ioctls per node, and then you can use one common function for all 3 s_fmt functions, calling the common try_fmt.